lxd/container_lxc.go

```diff
@@ -4006,21 +4145,108 @@ func (c *containerLXC) Update(args db.ContainerArgs, userRequested bool) error {
 			}
 		}
 
+		// wipe all per-port infiniband unix devices
+		dents, err := ioutil.ReadDir(c.DevicesPath())
+		if err != nil {
+			return err
+		}
+
+		for devName, m := range c.expandedDevices {
+			if m["type"] != "infiniband" {
+				continue
+			}
+
+			ifDevPrefix := fmt.Sprintf("infiniband.unix.%s.", devName)
+			for _, f := range dents {
+				// we only care about regular files that indicate a dependency on a per-port device
+				if !strings.HasPrefix(f.Name(), ifDevPrefix) || !f.Mode().IsRegular() {
+					continue
+				}
+
+				cmp := strings.Replace(f.Name(), fmt.Sprintf(".%s.", devName), ".", 1)
+				v, ok := infinibandUnixDevices[cmp]
+				if ok && !v {
+					infinibandUnixDevices[cmp] = true
+					break
+				}
+			}
+		}
+
+		for devName, v := range infinibandUnixDevices {
+			if v {
+				continue
+			}
+
+			tmp := strings.TrimPrefix(devName, "infiniband.unix.")
+			idx := strings.Index(tmp, ".")
+			if idx == -1 {
+				return fmt.Errorf("Invalid infiniband device \"%s\" detected", devName)
+			}
+			funName := tmp[:idx]
+
+			idx = strings.LastIndex(devName, ".")
+			if idx == -1 {
+				return fmt.Errorf("Invalid device name \"%s\"", devName)
+			}
+			relativeDestPath := devName[idx+1:]
+			relativeDestPath = strings.Replace(relativeDestPath, "-", "/", -1)
+			destPath := fmt.Sprintf("/%s", relativeDestPath)
+			dummyDevice := types.Device{
+				"path": destPath,
+			}
+			err = c.removeUnixDevice(fmt.Sprintf("infiniband.unix.%s", funName), dummyDevice)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Am I reading this right that every time anything is updated in the config of a container with an infiniband device, you're removing them all and re-adding them? Doesn't that sound like a terrible idea which will break running workloads?

Reply:
No, I'm only removing devices that need to be removed.

Reply:
Right, took me a while to track down, the mix of the infiniband (global) and infinibandUnixDevices (local) maps makes it pretty hard to track down.

Reply:
Ok, finally figured out what's going on here. It's because of the mix of per-port and per-device files, so in the event where we pass both ports into the container we don't want to end up removing it until both ports are gone.

I wonder if we couldn't simplify that whole mess by just passing the per-interface file twice in this case.
So say you have ib0 and ib1 in your devices, you'd end up with unix.ib0.dev-infiniband-uverbs0 and unix.ib1.dev-infiniband-uverbs1 and then handle the situation in insertUnixDevice and removeUnixDevice so that they become no-op if another unix.*.PATH already exists.

This would also fix this case:

    lxc config device add c1 kvm path=/dev/kvm
    lxc config device add c1 kvm1 path=/dev/kvm
    lxc config device remove c1 kvm

By effectively just dealing nicely with a user specifying the same device twice, first one wins and we don't remove until they're all gone.

Thoughts?

Reply:
> Right, took me a while to track down, the mix of the infiniband (global) and infinibandUnixDevices (local) maps makes it pretty hard to track down.

The commit message specifies in great detail how that is supposed to work and why it was done this way. :)

> I wonder if we couldn't simplify that whole mess by just passing the per-interface file twice in this case.
So say you have ib0 and ib1 in your devices, you'd end up with unix.ib0.dev-infiniband-uverbs0 and unix.ib1.dev-infiniband-uverbs1 and then handle the situation in insertUnixDevice and removeUnixDevice so that they become no-op if another unix.*.PATH already exists.

How do you handle the case where you remove the device that first added that per-port device that is mounted into the container? Everytime you remove a unix device you do a setns() into the container to check whether the device is mounted? If it is leave it on-disk and mounted if it isn't wipe it?. If that's the case then you still have to do the removal logic that we have here since you need to identify any stale on-disk per-port devices each time you remove any infiniband device. But now it becomes less nice since you don't now the device of the device anymore that first created the per-port on-disk device since you wiped it from the database. It's also not a true representation of the per-port concept on-disk since you pretend it's per-port but you still treat it as per-fun.

> This would also fix this case:
> 
> lxc config device add c1 kvm path=/dev/kvm
> lxc config device add c1 kvm1 path=/dev/kvm
> lxc config device remove c1 kvm
> By effectively just dealing nicely with a user specifying the same device twice, first one wins and we don't remove until they're all gone.

I don't understand we always failed on this and I don't see why mounting the same device to the same path should be valid.